File: mRemoteNG/Tools/ReconnectGroup.vb
﻿Public Class ReconnectGroup
    Private _ServerReady As Boolean
    Public Property ServerReady() As Boolean
        Get
            Return _ServerReady
        End Get
        Set(ByVal value As Boolean)
            If value = True Then
                SetStatusImage(My.Resources.HostStatus_On)
            Else
                SetStatusImage(My.Resources.HostStatus_Off)
            End If

            _ServerReady = value
        End Set
    End Property

    Private Delegate Sub SetStatusImageCB(ByVal Img As Image)
    Private Sub SetStatusImage(ByVal Img As Image)
        If pbServerStatus.InvokeRequired Then
            Dim d As New SetStatusImageCB(AddressOf SetStatusImage)
            ParentForm.Invoke(d, New Object() {Img})
        Else
            pbServerStatus.Image = Img
        End If
    End Sub

    Private Sub chkReconnectWhenReady_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles chkReconnectWhenReady.CheckedChanged
        _ReconnectWhenReady = chkReconnectWhenReady.Checked
    End Sub

    Private _ReconnectWhenReady As Boolean
    Public Property ReconnectWhenReady() As Boolean
        Get
            Return _ReconnectWhenReady
        End Get
        Set(ByVal value As Boolean)
            _ReconnectWhenReady = value
            SetCheckbox(value)
        End Set
    End Property

    Private Delegate Sub SetCheckboxCB(ByVal Val As Boolean)
    Private Sub SetCheckbox(ByVal Val As Boolean)
        If chkReconnectWhenReady.InvokeRequired Then
            Dim d As New SetCheckboxCB(AddressOf SetCheckbox)
            ParentForm.Invoke(d, New Object() {Val})
        Else
            chkReconnectWhenReady.Checked = Val
        End If
    End Sub

    Public Event CloseClicked()

    Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        RaiseEvent CloseClicked()
    End Sub

    Private Sub tmrAnimation_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles tmrAnimation.Tick
        Select Case lblAnimation.Text
            Case ""
                lblAnimation.Text = "»"
            Case "»"
                lblAnimation.Text = "»»"
            Case "»»"
                lblAnimation.Text = "»»»"
            Case "»»»"
                lblAnimation.Text = ""
        End Select
    End Sub

    Private Delegate Sub DisposeReconnectGroupCB()
    Public Sub DisposeReconnectGroup()
        If Me.InvokeRequired Then
            Dim d As New DisposeReconnectGroupCB(AddressOf DisposeReconnectGroup)
            ParentForm.Invoke(d)
        Else
            Me.Dispose()
        End If
    End Sub

    Private Sub ReconnectGroup_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
        ApplyLanguage()
    End Sub

    Private Sub ApplyLanguage()
        grpAutomaticReconnect.Text = My.Language.strGroupboxAutomaticReconnect
        btnClose.Text = My.Language.strButtonClose
        lblServerStatus.Text = My.Language.strLabelServerStatus
        chkReconnectWhenReady.Text = My.Language.strCheckboxReconnectWhenReady
    End Sub
End Class

File: mRemoteNG/Tools/ReconnectGroup.cs
using System;
using System.Drawing;
using Google.Protobuf;
using mRemoteNG.Resources.Language;

namespace mRemoteNG.Tools
{
    public partial class ReconnectGroup
    {
        public ReconnectGroup()
        {
            InitializeComponent();
            chkReconnectWhenReady.Checked = Properties.OptionsAdvancedPage.Default.NoReconnect;
        }

        private bool _ServerReady;

        public bool ServerReady
        {
            get => _ServerReady;
            set
            {
                SetStatusImage(value ? Properties.Resources.HostStatus_On : Properties.Resources.HostStatus_Off);

                _ServerReady = value;
            }
        }

        private delegate void SetStatusImageCB(Image Img);

        private void SetStatusImage(Image Img)
        {
            if (pbServerStatus.InvokeRequired)
            {
                var d = new SetStatusImageCB(SetStatusImage);
                ParentForm?.Invoke(d, new object[] {Img});
            }
            else
            {
                pbServerStatus.Image = Img;
            }
        }

        private void chkReconnectWhenReady_CheckedChanged(object sender, EventArgs e)
        {
            _ReconnectWhenReady = chkReconnectWhenReady.Checked;
        }

        private bool _ReconnectWhenReady;

        public bool ReconnectWhenReady
        {
            get => _ReconnectWhenReady;
            set
            {
                _ReconnectWhenReady = value;
                SetCheckbox(value);
            }
        }

        private delegate void SetCheckboxCB(bool Val);

        private void SetCheckbox(bool Val)
        {
            if (chkReconnectWhenReady.InvokeRequired)
            {
                var d = new SetCheckboxCB(SetCheckbox);
                ParentForm?.Invoke(d, new object[] {Val});
            }
            else
            {
                chkReconnectWhenReady.Checked = Val;
            }
        }

        public delegate void CloseClickedEventHandler();

        private CloseClickedEventHandler CloseClickedEvent;

        public event CloseClickedEventHandler CloseClicked
        {
            add => CloseClickedEvent = (CloseClickedEventHandler)Delegate.Combine(CloseClickedEvent, value);
            remove => CloseClickedEvent = (CloseClickedEventHandler)Delegate.Remove(CloseClickedEvent, value);
        }


        private void btnClose_Click(object sender, EventArgs e)
        {
            CloseClickedEvent?.Invoke();
        }

        private void tmrAnimation_Tick(object sender, EventArgs e)
        {
            switch (lblAnimation.Text)
            {
                case "":
                    lblAnimation.Text = "»";
                    break;
                case "»":
                    lblAnimation.Text = "»»";
                    break;
                case "»»":
                    lblAnimation.Text = "»»»";
                    break;
                case "»»»":
                    lblAnimation.Text = "";
                    break;
            }
        }

        private delegate void DisposeReconnectGroupCB();

        public void DisposeReconnectGroup()
        {
            if (InvokeRequired)
            {
                var d = new DisposeReconnectGroupCB(DisposeReconnectGroup);
                ParentForm?.Invoke(d);
            }
            else
            {
                Dispose();
            }
        }

        public void ReconnectGroup_Load(object sender, EventArgs e)
        {
            ApplyLanguage();
        }

        private void ApplyLanguage()
        {
            grpAutomaticReconnect.Text = Language.GroupboxAutomaticReconnect;
            btnClose.Text = Language._Close;
            lblServerStatus.Text = Language.ServerStatus;
            chkReconnectWhenReady.Text = Language.CheckboxReconnectWhenReady;
        }
    }
}